File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.py
from collections import deque

def solution(queue1, queue2):
    # 전체 요소 
    total_queue = queue1 + queue2
    # 속도를 위해 deque로 데이터 형태 변환
    queue1, queue2 = deque(queue1), deque(queue2)
    # 하나의 값인 경우 바로 return -1
    # 둘의 값이 홀수면 소용 없으니까 바로 return -1을 함
    if (sum(total_queue) % 2 == 1):
        return -1
    # 둘의 값이 같아야 하므로 반을 구함
    half = (sum(queue1) + sum(queue2)) // 2
    cnt = 0
    first_queue1 = sum(queue1)
    first_queue2 = sum(queue2)
    switch_length = len(total_queue)
    # 같을 때 까지 돌리기 
    while first_queue1 != first_queue2:
        if first_queue1 > first_queue2:
            value = queue1.popleft()
            queue2.append(value)
            first_queue1 -= value
            first_queue2 += value
        else:
            value = queue2.popleft()
            queue1.append(value)
            first_queue2 -= value
            first_queue1 += value
        # 처음으로 돌아가면 break하고 끝내버리기
        if cnt >= switch_length * 2:
            cnt = -1
            break
        cnt += 1
    return cnt

File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        int n = queue1.length;
        int[] arr = new int[2 * n];
        int left = 0;
        int right = n - 1;
        long total = 0;
        long sum = 0;  // queue1 합계
        int max = 0;
        for (int i = 0; i < n; i++) {
            arr[i] = queue1[i];
            arr[i + n] = queue2[i];
            sum += queue1[i];
            total += queue1[i] + queue2[i];
            max = Math.max(max, Math.max(queue1[i], queue2[i]));
        }
        if (total % 2 == 1) return -1;  // 애초에 둘 합해서 짝수가 아니면 못 구함
        long target = total / 2;  // 타겟
        // 
        if (max > target) return -1;  // 한 값이 타겟을 넘으면 불가
        int CAP = 3 * n;
        int moves = 0;
        while (moves <= CAP && right < 2 * n) {
            if (sum == target) {
                return moves;
            } else if (sum < target) {
                if (right + 1 == 2 * n) {
                    break;
                }
                sum += arr[++right];
            } else {
                sum -= arr[left++];
            }
            moves++;
        }
        return -1;
    }
}